File: apps/traverse-starter/windows-winui/TraverseStarter.Tests/PresentationMapperTests.cs
using System.Text.Json;
using Xunit;

namespace TraverseStarter.Tests;

public class PresentationMapperTests
{
    [Fact]
    public void EmptyStreamIsIdle()
    {
        var snap = PresentationMapper.MapPresentationState(Array.Empty<EmbedderEventLike>());
        Assert.Equal(PresentationState.Idle, snap.State);
        Assert.Null(snap.ErrorMessage);
    }

    [Fact]
    public void HappyPathLoads()
    {
        var events = new[]
        {
            new EmbedderEventLike(
                "capability_invoked",
                1,
                JsonDocument.Parse("""{"capability_id":"fixture.process"}""").RootElement),
            new EmbedderEventLike(
                "capability_result",
                2,
                JsonDocument.Parse("""{"capability_id":"fixture.process","output":{"ok":true}}""").RootElement),
        };

        Assert.Equal(PresentationState.Loaded, PresentationMapper.MapPresentationState(events).State);
        Assert.Null(PresentationMapper.ActiveCapabilityId(events));
        Assert.Equal(
            new[] { "fixture.process", "fixture.process" },
            PresentationMapper.MapCapabilityProgress(events).Select(s => s.CapabilityId).ToArray());
    }

    [Fact]
    public void BlockedWaitingForHuman()
    {
        var events = new[]
        {
            new EmbedderEventLike(
                "capability_invoked",
                1,
                JsonDocument.Parse("""{"capability_id":"fixture.approve"}""").RootElement),
            new EmbedderEventLike(
                "state_changed",
                2,
                JsonDocument.Parse("""{"state":"waiting_for_human"}""").RootElement),
        };

        Assert.Equal(PresentationState.Blocked, PresentationMapper.MapPresentationState(events).State);
        Assert.Equal("fixture.approve", PresentationMapper.ActiveCapabilityId(events));
    }
}

File: apps/loop/LoopCore/Tests/LoopCoreTests/PresentationMapperTests.swift
import XCTest
@testable import LoopCore

final class PresentationMapperTests: XCTestCase {
    func testEmptyStreamIsIdle() {
        let snap = PresentationMapper.mapPresentationState([])
        XCTAssertEqual(snap.state, .idle)
        XCTAssertNil(snap.errorMessage)
    }

    func testHappyPathLoads() {
        let events: [EmbedderEventLike] = [
            .init(eventType: "capability_invoked", sequence: 1, data: [
                "capability_id": "fixture.process",
            ]),
            .init(eventType: "capability_result", sequence: 2, data: [
                "capability_id": "fixture.process",
                "output": ["ok": true],
            ]),
        ]
        let snap = PresentationMapper.mapPresentationState(events)
        XCTAssertEqual(snap.state, .loaded)
        let progress = PresentationMapper.mapCapabilityProgress(events)
        XCTAssertEqual(progress.map(\.capabilityId), ["fixture.process", "fixture.process"])
        XCTAssertEqual(progress.map(\.phase), [.invoked, .result])
        XCTAssertNil(PresentationMapper.activeCapabilityId(events))
    }

    func testBlockedWaitingForHuman() {
        let events: [EmbedderEventLike] = [
            .init(eventType: "capability_invoked", sequence: 1, data: [
                "capability_id": "fixture.approve",
            ]),
            .init(eventType: "state_changed", sequence: 2, data: [
                "state": "waiting_for_human",
            ]),
        ]
        XCTAssertEqual(PresentationMapper.mapPresentationState(events).state, .blocked)
        XCTAssertEqual(PresentationMapper.activeCapabilityId(events), "fixture.approve")
    }

    func testErrorPath() {
        let events: [EmbedderEventLike] = [
            .init(eventType: "error", sequence: 1, data: [
                "error": ["message": "boom"],
            ]),
        ]
        let snap = PresentationMapper.mapPresentationState(events)
        XCTAssertEqual(snap.state, .error)
        XCTAssertEqual(snap.errorMessage, "boom")
    }
}
